File: src/BinaryTreeAsArray.java
import types.UserType;
import java.util.Arrays;


public class BinaryTreeAsArray {
    private UserType[] arr;
    private int size; //размер массива
    private int level;

    BinaryTreeAsArray() {
        init();
    }

    void init() {
        size = 1;
        level = 0;
        arr = new UserType[size];
    }

    BinaryTreeAsArray(UserType[] array) {
        level = calculateLevelFromSize(array.length);
        size = calculateSizeFromLevel(level);
        arr = Arrays.copyOf(array, size);
    }


    void insertByIndex(int n, UserType element) {
        if (element == null) return;

        if (n > size) {
            level++;
            size += calculateSizeOfLevel(level);
            arr = Arrays.copyOf(arr, size);
        }

        if (arr[n] == null) {
            arr[n] = element;
            return;
        }

        if (element.getTypeComparator().compare(element, arr[n]) > 0)
            insertByIndex(2*n + 2, element); //право
        else
            insertByIndex(2*n + 1, element); //влево
    }

    UserType getByIndex(int n) {
        return arr[n];
    }

    //балансировка
    void balance(UserType[] array, int a, int b) {
        if (a>b) return;
        int m = (a + b) / 2;
        insertByIndex(0, array[m]);
        balance(array, a, m-1);
        balance(array, m+1, b);
    }

    void set(UserType[] array, int n, int ln) {
        if (n>=size || arr[n]==null) return;
        set(array,2*n + 1,ln);
        array[ln++]=arr[n];
        set(array,2*n + 2,ln);
    }


    int sizer(int n) {//подсчет потомков
        if (n>=size || arr[n]==null) return 0;
        return 1 + sizer(2*n)+sizer(2*n+1);
    }

    void balance() {
        int  size1 = sizer(1);
        UserType[] tmp = Arrays.copyOf(arr, arr.length);
        init();
        balance(tmp, 0, size1);
    }

    //удаление
    void deleteByIndex(int index) {
        if (arr[index] == null) return;
        System.out.print("\n\ndeleted "+arr[index].readValue().toString() + " at "+index);
        delete(0,arr[index]);
    }

    int min(int rootIndex) {
        int ind = rootIndex;
        while (hasLeftChildren(ind)) {
            ind = 2*ind+1;
        }
        return ind;
    }

    int searchNextAfterRemoving(int ind) {
        if (hasRightChildren(ind)) {
            return min(2*ind+2);
        }
        int parInd = getParentIndex(ind);
        while (hasLeftChildren(parInd) && (arr[ind] == arr[2*parInd+2])) {
            arr[ind] = arr[parInd];
            parInd = getParentIndex(parInd);
        }
        return parInd;
    }


    int getParentIndex(int index) {
        if (index%2 == 0) return (index-2)/2;
         else return (index-1)/2;
    }


    UserType delete(int root, UserType element) {
        if (2*root+2>=size) return null;
        if (arr[root] == null) return arr[root];


        if (element.getTypeComparator().compare(element, arr[root]) < 0) { //если элемент < зн. узла = к левому потомку
            arr[2*root + 1] = delete(2*root + 1, element);
            return arr[root];
        }

        if (element.getTypeComparator().compare(element, arr[root]) > 0) {//если элемент > зн. узла = к правому потомку
            arr[2 * root + 2] = delete(2 * root + 2, element);
            return arr[root];
        }
        // если element найден

        if (arr[2*root+1] == null) {
            int temp = 2*root+2;
            arr[root] = null;
            return arr[temp];
        }
        if (arr[2*root+2] == null) {
            int temp = 2*root+1;
            arr[root] = null;
            return arr[temp];
        }

        arr[2*root+2] = deleteHelper(2*root+2,root); //если есть оба потомка
        return arr[root];
    }

    UserType deleteHelper(int root, int root0) {
        //if (2*root+2>=size) return null;
        if (2*root+2<size) {
            if (arr[2*root+1] != null) {
                arr[2*root+1] = deleteHelper(2*root+1, root0);
                return arr[root];
            }
        }

        arr[root0] = arr[root];
        arr[root] = null;
        //int temp = 2*root + 2;
       // arr[root] = null;
        if (2*root+2<size)  return arr[2*root + 2];
        else return null;
    }

    boolean hasRightChildren(int index) {
        if (2*index+2>=size) return false;
        return arr[2 * index + 2] != null;
    }

    boolean hasLeftChildren(int index) {
        if (2*index+2>=size) return false;
        return arr[2 * index + 1] != null;
    }

    void show() {
        System.out.println();
        for (int i = 0, cnt = 0, lvl = 0; i < size; i++) {
            if (i == cnt) {
                cnt += Math.pow(2, lvl);
                lvl++;
                System.out.println();
            }

            if (arr[i] == null) {
                System.out.print("N ");
                continue;
            }
            System.out.print(arr[i].readValue() + " ");
        }
    }

    private int calculateLevelFromSize(int sz) {
        int sum = 0;
        int level = -1;
        while (sum < sz) {
            level++;
            sum += Math.pow(2, level);
        }
        return level;
    }

    private int calculateSizeFromLevel(int level) {
        int sz = 0;
        for (int i = 0; i <= level; i++) {
            sz += Math.pow(2, i);
        }
        return sz;
    }

    private int calculateSizeOfLevel(int level) {
        return (int) Math.pow(2, level);
    }

    private boolean hasParent(int pos) {
        if (pos%2 == 0 && arr[pos/2] != null) return true;
        if (pos%2 != 0 && arr[(pos-1)/2] != null) return true;
        return false;
    }

    public void forEach(DoWith action) {
        for (int i=0; i<arr.length;i++) {
            String str;
            if (arr[i] == null) str = "null ";
            else str = arr[i].readValue().toString() + " ";
            action.doWith(str);
        }
    }

}

File: src/kotlin_module/BinaryTreeAsArray.kt
package kotlin_module

import java_module.BinaryTree
import java_module.DoWith
import java_module.types.UserType

class BinaryTreeAsArray() : BinaryTree {
    var size = 1 //размер массив
    private var arr: Array<UserType?> = arrayOfNulls(size)
    private var level = 0

    override fun init() {
        size = 1
        level = 0
        arr = arrayOfNulls(size)
    }

    internal constructor(array: Array<UserType>) : this() {
        level = calculateLevelFromSize(array.size)
        size = calculateSizeFromLevel(level)
        arr = array.copyOf(size)
    }

    private fun sizerLSS(n: Int): Int {
        return if (n >= size || arr[n] == null) 0 else 1 + sizerLSS(2 * n + 1) + sizerLSS(2 * n + 2)
    }

    override fun getByIndex(m: Int, n: Int): UserType? {
        var m = m
        if (m < 0 || m >= size || m >= sizerLSS(n)) return null
        val ll = sizerLSS(2 * n + 1)
        if (m < ll) return getByIndex(m, 2 * n + 1)
        m -= ll
        return if (m-- == 0) arr[n] else getByIndex(m, 2 * n + 2)
    }

    override fun insertByIndex(n: Int, element: UserType?) {
        if (element == null) return
        if (n >= size) {
            level++
            size += calculateSizeOfLevel(level)
            arr = arr.copyOf(size)
        }
        if (arr[n] == null) {
            arr[n] = element
            return
        }
        if (element.typeComparator.compare(element, arr[n]) > 0) insertByIndex(2 * n + 2, element) //право
        else insertByIndex(2 * n + 1, element) //влево
    }

    //балансировка
    private fun balance(array: ArrayList<UserType?>, a: Int, b: Int) {
        if (a > b) return
        val m = (a + b) / 2
        insertByIndex(0, array[m])
        balance(array, a, m - 1)
        balance(array, m + 1, b)
    }

    fun sizer(n: Int, arrayList: ArrayList<UserType?>): Int { //подсчет потомков
        if (n >= size || arr[n] == null) return 0
        arrayList.add(arr[n])
        return 1 + sizer(2 * n + 1, arrayList) + sizer(2 * n + 2, arrayList)
    }

    override fun balance() {
        val list = ArrayList<UserType?>()
        val size1 = sizer(0, list)
        list.sortWith(list[0]!!.typeComparator)
        init()
        balance(list, 0, size1 - 1)
    }

    //удаление
    override fun deleteByIndex(index: Int) {
        if (arr[index] == null) return
        delete(0, arr[index])
    }

    private fun delete(root: Int, element: UserType?): UserType? {
        if (2 * root + 2 >= size) return null
        if (arr[root] == null) return arr[root]
        if (element?.typeComparator!!.compare(element, arr[root]) < 0) { //если элемент < зн. узла = к левому потомку
            arr[2 * root + 1] = delete(2 * root + 1, element)!!
            return arr[root]
        }
        if (element.typeComparator!!.compare(element, arr[root]) > 0) { //если элемент > зн. узла = к правому потомку
            arr[2 * root + 2] = delete(2 * root + 2, element)!!
            return arr[root]
        }
        // если element найден
        if (arr[2 * root + 1] == null) {
            val temp = 2 * root + 2
            arr[root] = null
            return arr[temp]
        }
        if (arr[2 * root + 2] == null) {
            val temp = 2 * root + 1
            arr[root] = null
            return arr[temp]
        }
        arr[2 * root + 2] = deleteHelper(2 * root + 2, root)!! //если есть оба потомка
        return arr[root]
    }

    private fun deleteHelper(root: Int, root0: Int): UserType? {
        if (2 * root + 2 < size) {
            if (arr[2 * root + 1] != null) {
                arr[2 * root + 1] = deleteHelper(2 * root + 1, root0)!!
                return arr[root]
            }
        }
        arr[root0] = arr[root]
        arr[root] = null
        return if (2 * root + 2 < size) arr[2 * root + 2] else null
    }

    override fun show() {
        var i = 0
        var cnt = 0
        var lvl = 0
        while (i < size) {
            if (i == cnt) {
                cnt += Math.pow(2.0, lvl.toDouble()).toInt()
                lvl++
                println()
            }
            if (arr[i] == null) {
                print("N ")
                i++
                continue
            }
            print(arr[i].toString() + " ")
            i++
        }
    }

    override fun toString(): String {
        var str = ""
        var i = 0
        var cnt = 0
        var lvl = 0
        while (i < size) {
            if (i == cnt) {
                cnt += Math.pow(2.0, lvl.toDouble()).toInt()
                lvl++
                str += "\n"
            }
            if (arr[i] == null) {
                str += "N "
                i++
                continue
            }
            str += arr[i].toString() + " "
            i++
        }
        return str
    }

    private fun calculateLevelFromSize(sz: Int): Int {
        var sum = 0
        var level = -1
        while (sum < sz) {
            level++
            sum += Math.pow(2.0, level.toDouble()).toInt()
        }
        return level
    }

    private fun calculateSizeFromLevel(level: Int): Int {
        var sz = 0
        for (i in 0..level) {
            sz += Math.pow(2.0, i.toDouble()).toInt()
        }
        return sz
    }

    private fun calculateSizeOfLevel(level: Int): Int {
        return Math.pow(2.0, level.toDouble()).toInt()
    }

    override fun forEach(action: DoWith) {
        for (i in arr.indices) {
            val str: String = if (arr[i] == null) "null " else arr[i].toString() + " "
            action.doWith(str)
        }
    }
}